### PR TITLE
Initialize PipelineRun type and controller

### PR DESCRIPTION
**Pull request by JohnNiang**
### What this PR dose

- Initialize Jenkins PipelineRun API
- Create an empty PipelineRun controller

### Why we need it

This PR is part of #41 and is first step of the issue.

### CODE REVIEW

File: pkg/api/devops/v1alpha4/groupversion_info.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+package v1alpha4
+
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/scheme"
+)
+
+var (
+	// GroupVersion is group version used to register these objects
+	GroupVersion = schema.GroupVersion{Group: "devops.kubesphere.io", Version: "v1alpha4"}
```
Comment (LinuxSuRen) [inline pkg/api/devops/v1alpha4/groupversion_info.go:29]: How about creating a constant of these strings?

File: pkg/api/devops/v1alpha4/pipelinerun_types.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"kubesphere.io/devops/pkg/apis"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// PipelineRunSpec defines the desired state of PipelineRun
+type PipelineRunSpec struct {
+
```
Comment (LinuxSuRen) [inline pkg/api/devops/v1alpha4/pipelinerun_types.go:29]: Why keep an empty line at the first of the struct?

File: pkg/api/devops/v1alpha4/pipelinerun_types.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	// Parameters are some key/value pairs passed to runner.
+	// +optional
+	Parameters []*Parameter `json:"parameters,omitempty"`
+
+	// Scm is a SCM configuration that target pipeline run requires.
+	Scm *Scm `json:"scm,omitempty"`
```
Comment (LinuxSuRen) [inline pkg/api/devops/v1alpha4/pipelinerun_types.go:35]: `Scm` -> `SCM`

File: pkg/api/devops/v1alpha4/pipelinerun_types.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+	// Succeeded indicates that the pipeline run has succeeded.
+	Succeeded RunPhase = "Succeeded"
+
+	// Failed indicates that the pipeline run has failed.
+	Failed RunPhase = "Failed"
+
+	// Unknown indicates that the pipeline run has an unknown status.
+	Unknown RunPhase = "Unknown"
+)
```
Comment (LinuxSuRen) [inline pkg/api/devops/v1alpha4/pipelinerun_types.go:139]: These constants do not necessary in a versioned package. For example, we can put them in `pkg/api/utils.go`

File: pkg/api/devops/v1alpha4/pipelinerun_types.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+type ConditionType string
```
Comment (LinuxSuRen) [inline pkg/api/devops/v1alpha4/pipelinerun_types.go:141]: The comment is missing

**Review by LinuxSuRen — COMMENTED** [review]
I left some comments. Please take a look at it.

File: pkg/api/devops/v1alpha4/groupversion_info.go (hunk 1 above)
Comment (JohnNiang) [reply pkg/api/devops/v1alpha4/groupversion_info.go]: According to [Kubernetes apps API definition](https://github.com/kubernetes/api/blob/2491c6b734cf94a7efa5cb41a5a6c8ef2d4f8400/apps/v1beta1/register.go#L26-L30), I created only `GroupName` constant and kept `Version` string.

File: pkg/api/devops/v1alpha4/pipelinerun_types.go (hunk 2 above)
Comment (JohnNiang) [reply pkg/api/devops/v1alpha4/pipelinerun_types.go]: Delete all of them.

File: pkg/api/devops/v1alpha4/pipelinerun_types.go (hunk 3 above)
Comment (JohnNiang) [reply pkg/api/devops/v1alpha4/pipelinerun_types.go]: Changed all to SCM.

File: pkg/api/devops/v1alpha4/pipelinerun_types.go (hunk 4 above)
Comment (JohnNiang) [reply pkg/api/devops/v1alpha4/pipelinerun_types.go:133]: According to:
- https://github.com/kubernetes/api/blob/master/apps/v1/types.go
- https://github.com/kubernetes/api/blob/master/apps/v1beta1/types.go

I think those constants is reasonable in a versioned package.

File: pkg/api/devops/v1alpha4/pipelinerun_types.go (hunk 5 above)
Comment (JohnNiang) [reply pkg/api/devops/v1alpha4/pipelinerun_types.go:136]: Added.

File: pkg/api/devops/v1alpha4/groupversion_info.go (hunk 1 above)
Comment (LinuxSuRen) [reply pkg/api/devops/v1alpha4/groupversion_info.go]: Good example.

File: pkg/api/devops/v1alpha4/pipelinerun_types.go (hunk 4 above)
Comment (LinuxSuRen) [reply pkg/api/devops/v1alpha4/pipelinerun_types.go:133]: Fair enough.

**Review by LinuxSuRen — COMMENTED** [review]
/lgtm

**Pull request merged**